storage/innobase/fsp/fsp0fsp.cc

```
@@ -3577,22 +3571,401 @@ dberr_t fsp_sys_tablespace_validate()
+    if (err) return err;
+    err= add_index_root_pages(inodes, dict_sys.sys_fields);
+  }
+  return err;
+}
+
+/* Get the used inode from the system tablespace
+@param inodes inode information used found in system tablespace
+@param mtr    mini-transaction
+@return error code */
+static dberr_t fsp_get_sys_used_segment(inode_info *inodes, mtr_t *mtr)
+{
+  dberr_t err= DB_SUCCESS;
+  buf_block_t *block= nullptr;
+  /* Get TRX_SYS_FSEG_HEADER, TRX_SYS_DOUBLEWRITE_FSEG from
+  TRX_SYS_PAGE */
+  block= buf_pool.page_fix(page_id_t{0, TRX_SYS_PAGE_NO}, &err,
+                           buf_pool_t::FIX_WAIT_READ);
+  if (!block)
+    return err;
+
+  if (!inodes->insert_seg(block->page.frame + TRX_SYS + TRX_SYS_FSEG_HEADER))
+    err= DB_CORRUPTION;
+
+  if (!inodes->insert_seg(block->page.frame + TRX_SYS_DOUBLEWRITE
+                          + TRX_SYS_DOUBLEWRITE_FSEG))
+    err= DB_CORRUPTION;
+
+  block->page.unfix();
+
+  if (err)
+    return err;
+
+  block= buf_pool.page_fix(page_id_t{0, DICT_HDR_PAGE_NO}, &err,
+                           buf_pool_t::FIX_WAIT_READ);
+  if (!block)
+    return err;
+
+  if (!inodes->insert_seg(block->page.frame + DICT_HDR + DICT_HDR_FSEG_HEADER))
+    err= DB_CORRUPTION;
+
+  block->page.unfix();
+
+  if (err)
+    return err;
+
+  block= buf_pool.page_fix(page_id_t{0, FSP_IBUF_HEADER_PAGE_NO},
+                           &err, buf_pool_t::FIX_WAIT_READ);
+  if (!block)
+    return err;
+  if (!inodes->insert_seg(block->page.frame + PAGE_DATA))
+    err= DB_CORRUPTION;
+
+  block->page.unfix();
+
+  /* Get rollback segment header page */
+  for (ulint rseg_id= 0; rseg_id < TRX_SYS_N_RSEGS && err == DB_SUCCESS;
+       rseg_id++)
+  {
+    trx_rseg_t *rseg= &trx_sys.rseg_array[rseg_id];
+    if (rseg->space->id == 0)
+    {
+      block= buf_pool.page_fix(rseg->page_id(), &err,
+                               buf_pool_t::FIX_WAIT_READ);
+      if (!block)
+        continue;
```

Shouldn’t this be `break` or `return err` to avoid ignoring any error?